File: packages/gavio-py/gavio/request.py
"""GavioRequest — the canonical, provider-agnostic request model."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from ._ids import new_trace_id
from .types import Message, Provider


@dataclass
class GavioRequest:
    """A single gateway call.

    A ``trace_id`` (UUID v7, time-sortable) is assigned automatically if not
    supplied. ``parent_trace_id`` links calls into a multi-agent DAG.
    """

    messages: list[Message]
    model: str
    provider: Provider
    trace_id: str = field(default_factory=new_trace_id)
    agent_id: str | None = None
    parent_trace_id: str | None = None
    session_id: str | None = None
    options: dict[str, Any] = field(default_factory=dict)
    metadata: dict[str, Any] = field(default_factory=dict)

    def __post_init__(self) -> None:
        self.provider = Provider.coerce(self.provider)

    @property
    def temperature(self) -> float:
        return float(self.options.get("temperature", 0.7))

    @property
    def max_tokens(self) -> int:
        return int(self.options.get("max_tokens", 1024))

    def prompt_text(self) -> str:
        """Concatenate message contents — used for hashing and token estimation."""
        return "\n".join(m.get("content", "") for m in self.messages)

    def copy_with_messages(self, messages: list[Message]) -> GavioRequest:
        """Return a shallow copy with replaced messages (interceptors mutate via this)."""
        return GavioRequest(
            messages=messages,
            model=self.model,
            provider=self.provider,
            trace_id=self.trace_id,
            agent_id=self.agent_id,
            parent_trace_id=self.parent_trace_id,
            session_id=self.session_id,
            options=dict(self.options),
            metadata=dict(self.metadata),
        )

File: packages/gavio-js/src/request.ts
/** GavioRequest — the canonical, provider-agnostic request model. */

import { newTraceId } from './ids.js'
import { coerceProvider, PromptLineage } from './types.js'
import type { Message, PromptLineageInit, Provider } from './types.js'

export interface GavioRequestInit {
  messages: Message[]
  model: string
  provider: Provider | string
  traceId?: string
  agentId?: string | null
  parentTraceId?: string | null
  sessionId?: string | null
  options?: Record<string, unknown>
  metadata?: Record<string, unknown>
  /** Binary image inputs, scanned for PII before a provider call (F-SEC-09). */
  images?: Uint8Array[]
  lineage?: PromptLineage | PromptLineageInit | null
}

/**
 * A single gateway call. A `traceId` (UUID v7, time-sortable) is assigned
 * automatically if not supplied. `parentTraceId` links calls into a
 * multi-agent DAG.
 */
export class GavioRequest {
  messages: Message[]
  model: string
  provider: Provider
  traceId: string
  agentId: string | null
  parentTraceId: string | null
  sessionId: string | null
  options: Record<string, unknown>
  metadata: Record<string, unknown>
  images: Uint8Array[]
  lineage: PromptLineage | null

  constructor(init: GavioRequestInit) {
    this.messages = init.messages
    this.model = init.model
    this.provider = coerceProvider(init.provider)
    this.traceId = init.traceId ?? newTraceId()
    this.agentId = init.agentId ?? null
    this.parentTraceId = init.parentTraceId ?? null
    this.sessionId = init.sessionId ?? null
    this.options = init.options ?? {}
    this.metadata = init.metadata ?? {}
    this.images = init.images ?? []
    this.lineage = init.lineage != null ? PromptLineage.from(init.lineage) : null
  }

  get temperature(): number {
    const t = this.options['temperature']
    return typeof t === 'number' ? t : 0.7
  }

  get maxTokens(): number {
    const m = this.options['maxTokens'] ?? this.options['max_tokens']
    return typeof m === 'number' ? m : 1024
  }

  /** Concatenate message contents — used for hashing and token estimation. */
  promptText(): string {
    return this.messages.map((m) => m.content ?? '').join('\n')
  }

  /** Return a shallow copy with replaced messages (interceptors mutate via this). */
  copyWithMessages(messages: Message[]): GavioRequest {
    return new GavioRequest({
      messages,
      model: this.model,
      provider: this.provider,
      traceId: this.traceId,
      agentId: this.agentId,
      parentTraceId: this.parentTraceId,
      sessionId: this.sessionId,
      options: { ...this.options },
      metadata: { ...this.metadata },
      images: this.images,
      lineage: this.lineage,
    })
  }
}
